File: shop.c
//
//  shop.c
//  
//
//  Created by Isabella on 22/10/2021.
//

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

struct Product {
    char* name;
    double price;
};

struct ProductStock {
    struct Product product;
    int quantity;
};

struct Shop {
    double cash;
    struct ProductStock stock[20]; // Thats an array stock has 20 items
    int index;
};

struct Customer {
    char* name;
    double budget;
    struct ProductStock shoppingList[10];
    int index; // keeps track of value
};

void printProduct(struct Product p)
{
    printf("PRODUCT NAME: %s\nPRODUCT PRICE: €%.2f\n", p.name, p.price);
};

void printCustomer(struct Customer c)
{
    printf("--------------------------\n");
    printf("CUSTOMER NAME: %s\nCUSTOMER BUDGET: %.2f\n\n", c.name, c.budget);
    for (int i = 0; i< c.index; i++)
    {
        printf("--------------------------\n");
        printProduct(c.shoppingList[i].product);
        printf("QUANTITY: %d\n\n", c.shoppingList[i].quantity); // customer's shopping list, array item at index i, product quantity
        // cost = item quantity * price
        double cost = (c.shoppingList[i].quantity * c.shoppingList[i].product.price);
        printf("COST: €%.2f\n\n", cost);
    };
};

// Struct to create and stock shop
struct Shop createAndStockShop() 
{    
    FILE * fp;
    char * line = NULL;
    size_t len = 0;
    ssize_t read;

    // Opens csv with stock & shop value
    fp = fopen("csv_files/stock.csv", "r");
    
    if (fp == NULL)
        exit(EXIT_FAILURE);

    // Gets cash balance from csv
    read = getline(&line, &len, fp);
    float cash = atof(line);
    printf("CASH IN SHOP IS %.2f\n", cash);
    
    struct Shop shop = { cash };

    // String tokeniser method
    while ((read = getline(&line, &len, fp)) != -1) {
        // printf("Retrieved line of length %zu:\n", read);
        char *n = strtok(line, ",");
        char *p = strtok(NULL, ",");
        char *q = strtok(NULL, ",");
        double price = atof(p); // converts to floating point
        int quantity = atoi(q); // converts to integer
        char *name = malloc(sizeof(char) * 50); // Dynamcally allocated new memory for holding product name
        strcpy(name, n); // makes copy of data stored in n

        // Put data from CSV file into struct
        struct Product product = { name, price };
        struct ProductStock stockItem = { product, quantity };
        // adding items to shop
        shop.stock[shop.index++] = stockItem;
        // printf("NAME OF PRODUCT: %s\n PRICE: %.2f\n QUANTITY: %d\n\n", name, price, quantity);
    }

    return shop;

}
struct Customer customerOrderCSV()
{
    FILE * fp;
    char * line = NULL;
    size_t len = 0;
    ssize_t read;

    fp = fopen("customer1.csv", "r");
    if (fp == NULL)
        exit(EXIT_FAILURE);

    read = getline(&line, &len, fp);
    char *a = strtok(line, ",");
    char *b = strtok(NULL, ",");
    char *custName = malloc(sizeof(char) * 50);
    double custBudget = atof(b);
    strcpy(custName, a);
    struct Customer customer = { custName, custBudget }; 

    //struct ProductStock array[] = { StockA, StockB }
    while ((read = getline(&line, &len, fp)) != -1) {
        // code here
    }
    // return customer struct
    return customer;
}

void printShop(struct Shop *s) //added pointer
{
    printf("SHOP CASH FLOAT: €%.2f\n", s->cash);
    for (int i = 0; i < s->index; i++)
    {   
        printf("------------------------------\n");
        printProduct(s->stock[i].product);
        printf("QUANTITY AVAILABLE: %d\n", s->stock[i].quantity);
    }
}
// Display menu.
void displayMenu()
{
    // fflush(stdin); // might not need this
    printf("\n-------------------");
    printf("\nSHOP MENU\n");
    printf("-------------------");
    printf("\nView shop stock, press 1\n");
    printf("\nMake an order, press 2\n");
    printf("\nOption 3\n");
    printf("-------------------");
    printf("\nPress 0 to Exit\n");
    printf("-------------------\n");
}
// Return to menu.
void toMenu()
{   
    // Flush input.
    fflush(stdin); 
    char menu;
    printf("----------------------------\n");
    printf("To view menu press y.");
    scanf("%c", &menu);
    if (menu == 'y'){ // single quotes refers to value in a, while double refers to memory address
        displayMenu();
    }
};
double find(struct Shop s, char* name) 
{
    for (int i = 0; i < s.index; i++)
    {
        if (strcmp(name, s.stock[i].product.name) == 0){
            return s.stock[i].product.price;
        }
    }
    return -1;
}
// Main program.
int main(void)
{
    fflush(stdin);
    displayMenu();
    struct Shop shop = createAndStockShop();
    
    int choice = -1;
    while (choice != 0){
        printf("\nPlease select a menu option: ");
        scanf(" %d", &choice); // looks for input

        if (choice == 1){
            printf("Shop Stock & Status\n");
            // Prints stock items & shop balance
            printShop(&shop);
            toMenu();
        }
        else if (choice == 2){
            printf("Make an order\n");
            struct Customer customer = customerOrderCSV();
            // Check if customer has enough cash to purchase order
            
            if (customer.budget == 0) {
                printf("There is insufficient funds in your account.");
                toMenu();
            }
        }
    } //Put lengthy code in own method.
};

File: shop.py
from dataclasses import dataclass, field
from typing import List
import csv

@dataclass
class Product:
    name: str
    price: float = 0.0

@dataclass 
class ProductStock:
    product: Product
    quantity: int

@dataclass 
class Shop:
    cash: float = 0.0
    stock: List[ProductStock] = field(default_factory=list)

@dataclass
class Customer:
    name: str = ""
    budget: float = 0.0
    shopping_list: List[ProductStock] = field(default_factory=list) # creates empty list

# Populate shop with data from csv file.
def create_and_stock_shop():
    s = Shop()
    with open('csv_files/stock.csv') as csv_file:
        csv_reader = csv.reader(csv_file, delimiter=',')
        first_row = next(csv_reader)
        s.cash = float(first_row[0])
        for row in csv_reader:
            p = Product(row[0], float(row[1]))
            ps = ProductStock(p, float(row[2]))
            s.stock.append(ps)
            #print(ps)
    return s
   
def read_customer(file_path):
    with open(file_path) as csv_file:
        csv_reader = csv.reader(csv_file, delimiter=',')
        first_row = next(csv_reader)
        c = Customer(first_row[0], float(first_row[1]))
        for row in csv_reader:
            name = row[0]
            quantity = float(row[1])
            p = Product(name)           # Creates product
            ps = ProductStock(p, quantity)
            c.shopping_list.append(ps)
        return c 
        

def print_product(p):
    print(f'\nPRODUCT NAME: {p.name} \nPRODUCT PRICE: {p.price}')

def print_customer(c, s):
    print(f'CUSTOMER NAME: {c.name} \nCUSTOMER BUDGET: {c.budget}')
    total = 0
    for item in c.shopping_list:
        for p in s.stock:
            # Check if items exist.
            if item.product.name == p.product.name:
                print_product(p.product) # Prints each product in customer shopping list
                print(f'{c.name} ORDERS {item.quantity} OF ABOVE PRODUCT')
                cost = item.quantity * p.product.price
                print(f'THE COST TO {c.name} WILL BE €{cost}')
        total += cost
    print()
    print(f"TOTAL: {total}")
        
def print_shop(s):
    print(f'SHOP HAS {s.cash} IN CASH.')
    for item in s.stock:
        print_product(item.product)
        print(f'The Shop has {item.quantity} of the above')

def displayMenu():
    print()
    print("-------------------")
    print("Shop Menu")
    print("-------------------")
    print("Option 1")
    print("Option 2")
    print("Option 3")
    print("Press 0 to exit")

#s = create_and_stock_shop()
#print_shop(s)
if __name__ == '__main__':
    # displayMenu()
    # print("-------------------")
    # print()
    s = create_and_stock_shop()
    c = read_customer("csv_files/customer1.csv")
    print_customer(c, s)
    # print("----------------------------------")